cimi-resources/src/com/sixsq/slipstream/ssclj/resources/external_object/utils.clj

```diff
@@ -45,7 +51,109 @@
+      (log/error (format "Error when creating bucket %s: %s" bucketName (.getMessage e)))
+      false)))
+
+
+(defn authorized-bucket-operation?
+  [{:keys [objectStoreCred] :as resource} request check-authn-fn]
+  (-> objectStoreCred
+      (expand-cred)
+      (check-authn-fn request)))
+
+
+(defn uploadable-bucket?
+  "When the bucket exists, but the user can't create the object.
+  The external object resource must not be created."
+  [obj-store-conf bucketName]
+  (let [upload-url (generate-url obj-store-conf bucketName "test-probe" :put {:ttl 3 :content-type "text/plain" :filename "test.txt"})]
+    (= 200 (http/put upload-url))))
+
+
+(defn expand-obj-store-creds
+  "Need objectType to dispatch on when loading credentials."
+  [href-obj-store-cred]
+  (let [{:keys [key secret connector]} (expand-cred href-obj-store-cred)]
+    {:key      key
+     :secret   secret
+     :endpoint (:objectStoreEndpoint connector)}))
+
+
+(defn ok-to-add-external-resource?
+  "Determines if S3 conditions are met on S3 for the user to safely
+  add an external object resource. If everything is OK, then the resource
+  itself is returned. Otherwise an error response map is thrown"
+  [{:keys [bucketName objectStoreCred] :as resource} request]
+  (let [
+        obj-store-conf (expand-obj-store-creds objectStoreCred)
+        s3client (get-s3-client obj-store-conf)]
+    (cond
+      (not (bucket-creation-ok? s3client bucketName)) (logu/log-and-throw 503 (format "Unable to create the bucket %s" bucketName))
+      ;; When the requested bucket exists, but the user doesn't have permission to it :
+      ;; The external object resource must not be created."
+      (not (authorized-bucket-operation? resource request a/can-view?)) (logu/log-and-throw 403 (format "Access to bucket %s is not allowed" bucketName))
+      (not (uploadable-bucket? obj-store-conf bucketName)) (logu/log-and-throw 503 (format "Unable to create objects in the bucket %s" bucketName))
+      :all-ok resource)))
+
+
+(defn ok-to-delete-external-resource?
+  "Determines if S3 conditions are met on S3 for the user to safely
+  delete an external object resource. If everything is OK, then the resource
+  itself is returned. Otherwise an 'unauthorized' response map is thrown"
+  [{:keys [bucketName objectStoreCred] :as resource} request]
+  (let [
+        obj-store-conf (expand-obj-store-creds objectStoreCred)
+        s3client (get-s3-client obj-store-conf)]
+    (cond
+      ;; When the user requests to delete an object, but it no longer exists :
+      ;; The external object resource should be deleted normally.
+      (not (bucket-exists? s3client bucketName)) resource
+      ;; "When the user cannot access the bucket/object to delete it :
+      ;; the external object resource should not be deleted."
+      (not (authorized-bucket-operation? resource request a/can-modify?)) (logu/log-and-throw 403 (format "Deleting the bucket %s is not allowed" bucketName))
+      :all-ok resource)))
+
```

Comment:
Shouldn't this also check the case where the bucket exists but the object does not?

Reply:
done